### DISCUSSION TITLE
Weird diff behaviour

### DISCUSSION THREAD

**Discussion by RussKie** [gitextensions/gitextensions#9096] [Q&A]
Observed an puzzling behaviour - if I select two or more commits with <kbd>CTRL</kbd>+click I get one diff, and choosing the same commits in the same order by with <kbd>SHIFT</kbd>+click yields the opposite result.

E.g.
1. Select 885d886465c93446b488a1c17f87efab87b768b0 first, 
2. Select 240ff0b6f01672ec51bd38a34580a82f1ff7e631 second.

* <kbd>CTRL</kbd>+click
![image](https://user-images.githubusercontent.com/4403806/115543591-e4373a80-a2e4-11eb-9647-2331afcbf8e5.png)

* <kbd>SHIFT</kbd>+click
![image](https://user-images.githubusercontent.com/4403806/115543643-f1542980-a2e4-11eb-88d6-9b5021add891.png)

@gerhardol any thoughts?

**Comment by mstv** [comment]
![grafik](https://user-images.githubusercontent.com/36601201/115611209-14350c80-a2ea-11eb-9b22-5dddf22d2b24.png)

**Reply by mstv** [reply]
The sequence of clicks does matter.

**Reply by pmiossec** [reply]
@mstv I agree with @RussKie, that's strange and I don't understand why the 2 give the opposite result.

The 2 commits have been selected in the same order. Just one time with <kbd>CTRL</kbd> and the other with <kbd>SHIFT</kbd>. And don't see an explanation in the tooltip...

**Comment by vbjay** [comment]
Is the order based on commit order vs selected order when shift is used.
Shift allows many and not just one commit. So I bet order is calculated
from commit order.

**Comment by gerhardol** [comment]
The diff presentation is based on the order in GetRevisions.
If you ctrl-click, the clicked commit is placed first in the array, which is handled as it is selected last.
Shift-click seem to order the commits from oldest to newest instead.

The only reason I can think of is to cherry-pick a range of commits.
Shift-click could be disabled.

I have no reason to shift click and have not reacted on this myself.

**Comment by RussKie** [comment]
I shift-clicked automatically because the commits were next to each other.
We'll need to get it fixed, this is uber confusing.

**Reply by gerhardol** [reply]
> I shift-clicked automatically because the commits were next to each other. We'll need to get it fixed, this is uber confusing.

Remove shift or handle ordering?
I see no point in shift-click, can be useful...

**Reply by gerhardol** [reply]
I do not see any way to handle this in a DataGridView.
A kind of workaround below, but the code for selections is complex enough already...
https://github.com/gerhardol/gitextensions/pull/new/feature/i9096-shift-click-grid

**Reply by gerhardol** [reply]
@RussKie Any interest in the proposal?
I do not like it myself though...

**Reply by RussKie** [reply]
Hmm.. one day I'll probably get around looking in to it.

**Comment by vbjay** [comment]
Shift click is used in patch commands. It is on purpose. Shift click allows
you to say all commits between x and y commits. So order is controled by
commit order.

Ctrl click is I select x and then y then z then....

We just need to show something showing the affect of ctrl vs shift in
status bar.
